x-pack/plugins/endpoint/server/routes/resolver/utils/tree.ts

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import { ResolverEvent, Node, NodeStats, NodePagination } from '../../../../common/types';
+import { extractEntityID, extractParentEntityID } from './normalize';
+import { buildPaginationCursor } from './pagination';
+
+type ExtractFunction = (event: ResolverEvent) => string | undefined;
+
+function createNode(id: string): Node {
+  return { id, children: [], pagination: {}, events: [], alerts: [], lifecycle: [] };
+}
+
+// This class aids in constructing a tree of process events. It works in the following way:
+//
+// 1. We construct a tree structure starting with the root node for the event we're requesting.
+// 2. We leverage the ability to pass hashes and arrays by reference to construct a fast cache of
+//    process identifiers that updates the tree structure as we push values into the cache.
+//
+// When we query a single level of results for child process events we have a flattened, sorted result
+// list that we need to add into a constructed tree. We also need to signal in an API response whether
+// or not there are more child processes events that we have not yet retrieved, and, if so, for what parent
+// process. So, at the end of our tree construction we have a relational layout of the events with no
+// pagination information for the given parent nodes. In order to actually construct both the tree and
+// insert the pagination information we basically do the following:
+//
+// 1. Using a terms aggregation query, we return an approximate roll-up of the number of child process
```

Comment:
please justify why this is necessary

Reply:
I think this is so we can so totals in the UI.